salt/pillar/__init__.py

```diff
@@ -302,6 +298,119 @@ def __init__(self, opts, grains, minion_id, saltenv, ext=None, functions=None,
             else:
                 log.error('Pillar data must be a dictionary')
 
+    @property
+    def pillar(self):
+        '''
+        Return the current pillar state.
+        '''
+        # FIXME: should this return a deepcopy rather than a reference?
+        # FIXME: should this merge in self.pillar_override?
+        # FIXME: maybe keep a hash of the accumulator to check if other objects need to regenerate?
+        pillar = copy.deepcopy(self._pillar_accumulator)
```

Comment:
*Yes*.  Not only is ``self.pillar`` being generated *each* time, but other objects are being regenerated.  This goes along with the third ``FIXME`` above where I muse on the possibility of tracking when this and other related objects need to be regenerated - that way rereferencing ``self.pillar`` a second time would not incur a rebuild when nothing has changed.

All of that adds to the complexity, but at this point I'm thinking that guaranteeing up-to-date pillar from previous pillar sources is going to require non-trivial improvements to the ``Pillar`` class.